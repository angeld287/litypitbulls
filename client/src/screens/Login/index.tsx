import { Content } from 'antd/lib/layout/layout';
import { MessageApi } from 'antd/lib/message';
import React, { useCallback, useEffect, useState } from 'react';
import { useAppDispatch, useAppSelector } from '../../app/hooks';
import { ICustomButton } from '../../components/CustomButton/ICustomButton';
import CustomForm from '../../components/CustomForm';
import { ICustomFields } from '../../components/CustomForm/ICustomForm';
import { setIsRegistering } from '../../features/userRegister/userRegisterSlice';
import { loginAsync } from '../../features/userSession/asyncThunks';
import { ICredentials } from '../../features/userSession/IUserSession';
import { selectUserSession } from '../../features/userSession/userSessionSlice';
import { IValidationError } from '../../interfaces/models/IBase';
import styles from './styles';

const Login: React.FC = () => {

    const session = useAppSelector(selectUserSession);
    const dispatch = useAppDispatch()
    const [message, setMessage] = useState<MessageApi>()
    const [error, setError] = useState<IValidationError>()

    useEffect(() => {
        const login = () => {

            setError({
                message: '',
                error: false
            });

            if (!message)
                return

            let sessionError = session.error;
            if (sessionError) {
                if (!Array.isArray(sessionError)) {
                    setError({
                        message: sessionError.message,
                        error: true
                    })
                    return message.error(sessionError.message)
                }
                if (Array.isArray(sessionError)) {
                    setError({
                        message: sessionError[0].message,
                        error: true
                    })
                    return sessionError.forEach(e => {
                        message.error(e.message)
                    });
                }
            }

            message.success('Login Successfully!')
        }

        if (session.loginStatus === 'idle' && message) {
            login()
        } else if (session.loginStatus === 'failed') {
            setError({
                message: 'Internal Error',
                error: true
            })
        }

    }, [session.loginStatus, session.error, message]);

    const handleClicLoginButton = useCallback(
        (credentials: ICredentials, message: MessageApi) => {
            setMessage(message);
            dispatch(loginAsync(credentials))
        }
        , [dispatch]
    );

    const goToRegisterScreen = useCallback(
        () => {
            dispatch(setIsRegistering(true));
        },
        [dispatch]
    )

    let inputFields: Array<ICustomFields> = [
        {
            name: 'username',
            input: {
                name: 'username',
                label: 'Username',
                defaultValue: '',
                disabled: false,
                type: 'input',
            }
        },
        {
            name: 'password',
            input: {
                name: 'password',
                label: 'Password',
                defaultValue: '',
                disabled: false,
                type: 'password',
            }
        },
    ]

    let btns: Array<ICustomButton> = [
        {
            color: 'blue',
            _key: 'register_btn',
            children: 'Sign Up',
            htmlType: 'button',
            name: "register",
            onClick: goToRegisterScreen
        },
        {
            color: 'blue',
            _key: 'login_btn',
            children: 'Login',
            loading: session.loginStatus === 'pending',
            htmlType: 'submit',
            name: "login",
        },
    ]

    return (
        <>
            <Content style={styles.container}>
                <CustomForm error={error} onSubmit={handleClicLoginButton} fields={inputFields} buttons={btns} verticalButtons={false} loading={session.loginStatus === 'pending'} />
            </Content>
        </>
    );
};

export default Login;