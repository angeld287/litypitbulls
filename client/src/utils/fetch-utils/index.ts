export const fetcher = async (input: RequestInfo | URL, init?: RequestInit | undefined): Promise<any> => {
    const res = await fetch(input, init)
    return await res.json()
}