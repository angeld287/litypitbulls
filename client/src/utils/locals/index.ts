// @ts-nocheck
/**
 * Define App Locals & Configs
 *
 * @author Angel Angeles <aangeles@litystyles.com>
 */

class Locals {

    /**
     * Makes env configs available for your app
     * throughout the app's runtime
     */
    public static config(): any {
        const server_host = window._env_.REACT_APP_SERVER_HOST;
        const server_port = window._env_.REACT_APP_SERVER_PORT;

        const server_url = "http://" + server_host + ":" + server_port

        return {
            server_url
        }
    }
}

export default Locals;