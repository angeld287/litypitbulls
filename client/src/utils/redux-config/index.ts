import { configureStore } from "@reduxjs/toolkit";
import userSessionReducer from '../../features/userSession/userSessionSlice';
import userRegisterReducer from '../../features/userRegister/userRegisterSlice';

export const store = configureStore({
    reducer: {
        userSession: userSessionReducer,
        userRegister: userRegisterReducer

    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware({
        serializableCheck: false
    })
});