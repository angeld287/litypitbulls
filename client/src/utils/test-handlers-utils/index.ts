import { ResponseTransformer, rest, RestContext } from "msw";

export let json = (ctx: RestContext, obj: any): ResponseTransformer<any, any> => ctx.json(obj)

export const headers = (ctx: RestContext): Array<ResponseTransformer<any, any>> => {
    return [
        ctx.status(200),
        ctx.set("access-control-allow-origin", "*"),
        ctx.set("Accept", "application/json"),
        ctx.set("Content-Type", "application/json")
    ]
}

export const _rest = rest;