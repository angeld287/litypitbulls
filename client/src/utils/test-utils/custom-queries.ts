import { queryHelpers, buildQueries, Matcher, MatcherOptions } from '@testing-library/react'

// The queryAllByAttribute is a shortcut for attribute-based matchers
// You can also use document.querySelector or a combination of existing
// testing library utilities to find matching nodes for your query
const queryAllByDataCy = (container: HTMLElement, id: Matcher, options?: MatcherOptions | undefined) =>
    queryHelpers.queryAllByAttribute('data-cy', container, id, options)

const queryAllByAriaOwns = (container: HTMLElement, id: Matcher, options?: MatcherOptions | undefined) =>
    queryHelpers.queryAllByAttribute('aria-owns', container, id, options)

const queryAllByClass = (container: HTMLElement, id: Matcher, options?: MatcherOptions | undefined) =>
    queryHelpers.queryAllByAttribute('class', container, id, options)

const queryAllByType = (container: HTMLElement, id: Matcher, options?: MatcherOptions | undefined) =>
    queryHelpers.queryAllByAttribute('type', container, id, options)

const getMultipleError = (c: any, value: string) =>
    `Found multiple elements with the attribute: ${value}`
const getMissingError = (c: any, value: string) =>
    `Unable to find an element with the attribute: ${value}`

const [
    queryByDataCy,
    getAllByDataCy,
    getByDataCy,
    findAllByDataCy,
    findByDataCy,
] = buildQueries(queryAllByDataCy, getMultipleError, getMissingError)

const [
    queryByClass,
    getAllByClass,
    getByClass,
    findAllByClass,
    findByClass,
] = buildQueries(queryAllByClass, getMultipleError, getMissingError)

const [
    queryByAriaOwns,
    getAllByAriaOwns,
    getByAriaOwns,
    findAllByAriaOwns,
    findByAriaOwns,
] = buildQueries(queryAllByAriaOwns, getMultipleError, getMissingError)

const [
    queryByType,
    getAllByType,
    getByType,
    findAllByType,
    findByType,
] = buildQueries(queryAllByType, getMultipleError, getMissingError)

export {
    queryByDataCy,
    queryAllByDataCy,
    getByDataCy,
    getAllByDataCy,
    findAllByDataCy,
    findByDataCy,
    getAllByAriaOwns,
    queryByAriaOwns,
    getByAriaOwns,
    findAllByAriaOwns,
    findByAriaOwns,
    queryAllByAriaOwns,
    queryByClass,
    getAllByClass,
    getByClass,
    findAllByClass,
    findByClass,
    queryAllByClass,
    queryByType,
    getAllByType,
    getByType,
    findAllByType,
    findByType,
}