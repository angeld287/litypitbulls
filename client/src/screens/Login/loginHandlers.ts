import { headers, json, _rest } from "../../utils/test-handlers-utils";

const login = _rest.post('http://localhost:3001/api/auth/login', (req, res, ctx) => {

    const _headers = headers(ctx);
    let _json = json(ctx, {});

    let _result: any = null;

    if (!req.body)
        return res(..._headers, _json, ctx.delay(100))

    _result = { errors: [{ message: "", value: "", param: "username", location: "body" }] }

    let _body: any = req.body; //existingadmin@test.com

    if (_body.username === "")
        _result.errors[0].message = "E-mail cannot be blank."

    if (_body.username === "existingadmin")
        _result.errors[0].message = "E-mail is not valid."

    if (_body.password === "")
        _result.errors[0].message = "Password cannot be blank."

    if (_body.password === "admin")
        _result.errors[0].message = "Password length must be atleast 8 characters."

    if (_body.username === "existingadmin@test.com" && _body.password === "badPass")
        _result.errors[0].message = "Invalid Username or Password."

    if (_body.username === "existingadmin@test.com" && _body.password === "admin2807")
        _result = {
            session: {
                id: 4,
                email: "existingadmin@test.com",
                phoneNumber: "8095445501",
                passwordResetToken: null,
                passwordResetExpires: null,
                fullname: "Test User Name",
                gender: "m",
                profile: null,
                userName: "existingAdmin"
            }
        };

    _json = json(ctx, {
        statusCode: "10000", message: "Success",
        data: _result
    });

    return res(..._headers, _json, ctx.delay(10))
});

const handlers = [login]

export default handlers