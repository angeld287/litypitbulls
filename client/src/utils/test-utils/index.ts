// test-utils.js
import { render, queries, RenderOptions } from '@testing-library/react'
import { ReactElement } from 'react';
import * as customQueries from './custom-queries'
import * as functions from './functionalities';

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { queries: { ...queries, ...customQueries }, ...options })


// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render, functions }