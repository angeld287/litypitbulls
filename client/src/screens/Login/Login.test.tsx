import React from 'react';
import { render, RenderResult, functions, fireEvent, screen, waitFor, act } from '../../utils/test-utils';
import { Provider } from 'react-redux';
import { setupServer } from 'msw/node'
import handlers from './loginHandlers';

import Login from './index';
import { store } from '../../utils/redux-config';

let component: RenderResult;

const server = setupServer(...handlers)

describe("Login Test Suite", () => {
    beforeAll(() => server.listen({
        onUnhandledRequest: ({ headers, method, url }) => {
            if (headers.get("User-Agent") !== "supertest") {
                throw new Error(`Unhandled ${method} request to ${url}`);
            }
        },
    }));

    beforeEach(() => {
        if (component)
            component.unmount()

        act(() => {
            component = render(
                <Provider store={store}>
                    <Login />
                </Provider>
            )
        });
        server.resetHandlers()
    });

    afterAll(() => server.close())

    test('It must render the component correcly', () => {
        expect(component.asFragment()).toMatchSnapshot();
    })

    test('It must display basic login components', () => {
        expect(component.getByText(/Username/i)).toBeInTheDocument();
        expect(component.getByText(/Password/i)).toBeInTheDocument();
        expect(component.getByText(/Login/i)).toBeInTheDocument();
    });


    test('It must respond "E-mail cannot be blank." when email is blank.', async () => {

        functions.writeInInputFoundByPlaceHolder(null, /Username/i, "");
        functions.writeInInputFoundByPlaceHolder(null, /Password/i, "admin2807");

        await act(() => {
            fireEvent.click(screen.getByText(/Login/i));
        });

        await waitFor(() => {
            component.rerender(
                <Provider store={store}>
                    <Login />
                </Provider>
            )
            expect(screen.getByText("E-mail cannot be blank.")).toBeInTheDocument();
        });
    });

    test('It must respond "E-mail is not valid" when email is invalid.', async () => {

        functions.writeInInputFoundByPlaceHolder(null, /Username/i, "existingadmin");
        functions.writeInInputFoundByPlaceHolder(null, /Password/i, "admin2807");

        await act(() => {
            fireEvent.click(screen.getByText(/Login/i));
        });

        await waitFor(() => {
            component.rerender(
                <Provider store={store}>
                    <Login />
                </Provider>
            )
            expect(screen.getByText("E-mail is not valid.")).toBeInTheDocument();
        });
    });

    test('It must respond "Password cannot be blank" when password is blank.', async () => {

        functions.writeInInputFoundByPlaceHolder(null, /Username/i, "existingadmin@test.com");
        functions.writeInInputFoundByPlaceHolder(null, /Password/i, "");

        await act(() => {
            fireEvent.click(screen.getByText(/Login/i));
        });

        await waitFor(() => {
            //verify if validation message is shown
            expect(screen.getByText("Password cannot be blank.")).toBeInTheDocument();
        });
    });

    test('It must respond "Password length must be atleast 8 characters." when password is invalid', async () => {

        functions.writeInInputFoundByPlaceHolder(null, /Username/i, "existingadmin@test.com");
        functions.writeInInputFoundByPlaceHolder(null, /Password/i, "admin");

        await act(() => {
            fireEvent.click(screen.getByText(/Login/i));
        });

        await waitFor(() => {
            //verify if validation message is shown
            expect(screen.getByText("Password length must be atleast 8 characters.")).toBeInTheDocument();
        });
    });

    test('It must respond "Invalid Username or Password." when password is invalid', async () => {

        functions.writeInInputFoundByPlaceHolder(null, /Username/i, "existingadmin@test.com");
        functions.writeInInputFoundByPlaceHolder(null, /Password/i, "badPass");

        await act(() => {
            fireEvent.click(screen.getByText(/Login/i));
        });

        await waitFor(() => {
            //verify if validation message is shown
            expect(screen.getByText("Invalid Username or Password.")).toBeInTheDocument();
        });
    });
});