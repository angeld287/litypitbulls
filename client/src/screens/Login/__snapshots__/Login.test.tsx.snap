// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Test Suite It must render the component correcly 1`] = `
<DocumentFragment>
  <main
    class="ant-layout-content"
    style="width: 400px; height: 400px; margin: 50px auto; background-color: white; flex-direction: column; justify-content: center; align-items: center;"
  >
    <form
      autocomplete="off"
      class="ant-form ant-form-horizontal"
      id="basic"
    >
      <div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-8 ant-form-item-label"
          >
            <label
              class=""
              for="basic_username"
              title="Username"
            >
              Username
            </label>
          </div>
          <div
            class="ant-col ant-col-16 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="basic_username"
                  placeholder="Type your Username"
                  type="input"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-8 ant-form-item-label"
          >
            <label
              class=""
              for="basic_password"
              title="Password"
            >
              Password
            </label>
          </div>
          <div
            class="ant-col ant-col-16 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="basic_password"
                  placeholder="Type your Password"
                  type="password"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        style="text-align: right;"
      >
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-16 ant-col-offset-8 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn ant-btn-default"
                  style="color: blue;"
                  type="button"
                >
                  <span>
                    Sign Up
                  </span>
                </button>
                <button
                  class="ant-btn ant-btn-default"
                  style="color: blue;"
                  type="submit"
                >
                  <span>
                    Login
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </main>
</DocumentFragment>
`;
